File: grpc-client/src/test/python/util/google_cloud_storage_util.py
from google.cloud import storage
import os

bucket_name = "embedcraft"

def upload_model_to_gcs(client, destination_blob_name, model_bytes):
    """Uploads a file to the bucket."""
    
    bucket = client.bucket(bucket_name)

    blob = bucket.blob(destination_blob_name)
    blob.upload_from_string(model_bytes)


def read_file_from_gcs(client, blob_name):
    """Reads a file from Google Cloud Storage and returns its contents."""
    
    # Get the bucket and blob (file) from GCS
    bucket = client.bucket(bucket_name)
    blob = bucket.blob(blob_name)
    
    # Download the blob's contents as a string
    contents = blob.download_as_text()

    return contents

def initialize_client():
    # Authenticate to GCS using the service account key file
    return storage.Client.from_service_account_json("./util/service-account-file.json")
    

# if __name__ == "__main__":
#     client = initialize_client()
#     print(read_file_from_gcs(client, "dataset/35bc50b4-8794-4940-be73-d3307518eb71.txt"))

File: embedcraft-core/src/main/java/com/embedcraft/embedcraftcore/util/GoogleCloudStorageUtil.java
package com.embedcraft.embedcraftcore.util;

import com.google.auth.oauth2.GoogleCredentials;
import com.google.cloud.storage.BlobId;
import com.google.cloud.storage.BlobInfo;
import com.google.cloud.storage.Storage;
import com.google.cloud.storage.StorageOptions;
import org.springframework.core.io.ClassPathResource;

import java.io.IOException;
import java.util.List;
import java.util.UUID;

public class GoogleCloudStorageUtil {
    private static Storage storage;
    private final static String bucketName= "embedcraft";
    private static final String datasetFolderPath = "dataset/"; // dataset folder path within the bucket

    static {
        // Key file inside the resources folder
        GoogleCredentials credentials;
        try {
            credentials = GoogleCredentials.fromStream(new ClassPathResource("service-account-file.json").getInputStream())
                    .createScoped(List.of("https://www.googleapis.com/auth/cloud-platform"));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        // Initializes the GCS client
        storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService();
    }

    public static String uploadTextFile(byte[] content) {
        // Generate a random file name using UUID
        String objectName = datasetFolderPath + UUID.randomUUID() + ".txt";

        BlobId blobId = BlobId.of(bucketName, objectName);
        BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
        // Upload the file
        storage.create(blobInfo, content);

//         Alternatively, the blob's ID (bucketName/objectName) as the identifier:

        return blobId.getName(); // Return the URL or fileId
    }
}
